File: dart/lib/src/record/get_record_v2_address.dart
import '../constants/addresses.dart';
import '../constants/records.dart';
import '../domain/get_domain_address.dart';
import '../utils/derive_address.dart';

/// Parameters for getting a V2 record address
class GetRecordV2AddressParams {
  const GetRecordV2AddressParams({
    required this.domain,
    required this.record,
  });

  /// The domain under which the record resides
  final String domain;

  /// The type of record to derive the address for
  final Record record;
}

/// Derives the address of a version 2 record.
///
/// This function mirrors js-kit/src/record/getRecordV2Address.ts
///
/// [params] - Parameters containing the domain and record type
///
/// Returns the derived record address
Future<String> getRecordV2Address(GetRecordV2AddressParams params) async {
  final domainResult = await getDomainAddress(GetDomainAddressParams(
    domain: params.domain,
  ));

  return deriveAddress(
    '\x02${params.record.name}',
    parentAddress: domainResult.domainAddress,
    classAddress: centralStateDomainRecords,
  );
}

File: js-kit/src/record/getRecordV2Address.ts
import { Address } from "@solana/kit";

import { CENTRAL_STATE_DOMAIN_RECORDS } from "../constants/addresses";
import { getSnsDomainAddress } from "../domain/getSnsDomainAddress";
import { Record } from "../types/record";
import { _deriveAddress } from "../utils/deriveAddress";

/**
 * Parameters for deriving a V2 record address.
 *
 * @example
 * ```ts
 * const params: GetRecordV2AddressParams = { domain: "example", record: Record.Url };
 * ```
 */
export interface GetRecordV2AddressParams {
  /** TLD-less domain name. */
  domain: string;
  /** Record type. */
  record: Record;
}

/**
 * Derives the address of a V2 record account.
 *
 * @param params Record address derivation parameters
 * @param params.domain TLD-trimmed SNS domain name
 * @param params.record Record type
 * @returns The derived V2 record account address.
 *
 * @example
 * ```ts
 * const address = await getRecordV2Address({ domain: "example", record: Record.Url });
 * ```
 */
export const getRecordV2Address = async ({
  domain,
  record,
}: GetRecordV2AddressParams): Promise<Address> => {
  const { domainAddress } = await getSnsDomainAddress({ domain });

  return await _deriveAddress(
    `\x02${record}`,
    domainAddress,
    CENTRAL_STATE_DOMAIN_RECORDS
  );
};
